test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/ImmutableModelTest.java

```diff
@@ -59,4 +62,27 @@ public void mapsOfListsAreImmutable() {
         assertThatExceptionOfType(UnsupportedOperationException.class)
                 .isThrownBy(() -> request.mapOfStringToIntegerList().get("key").add(2));
     }
+
+    @Test
+    public void byteBuffersAreImmutable() {
+        ByteBuffer buffer = ByteBuffer.wrap("Hello".getBytes(StandardCharsets.UTF_8));
+
+        int position = buffer.position();
+        int limit = buffer.limit();
+        ByteOrder order = buffer.order();
+
+        AllTypesRequest request = AllTypesRequest.builder().blobArg(buffer).build();
+
+        buffer.limit(4);
+        buffer.position(0);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+        buffer.put("Foo".getBytes(StandardCharsets.UTF_8));
+
+        assertThat(request.blobArg()).as("Check new read-only blob each time").isNotSameAs(request.blobArg());
+        assertThat(request.blobArg().isReadOnly()).as("Check read-only").isTrue();
+        assertThat(request.blobArg().limit()).as("Check limit").isEqualTo(limit);
+        assertThat(request.blobArg().position()).as("Check position").isEqualTo(position);
+        assertThat(request.blobArg().order()).as("Check order").isEqualTo(order);
+        assertThat(request.blobArg()).as("Check concurrent modification").isNotEqualByComparingTo(buffer);
```

Comment:
Yeah, I suppose not. I thought of it as "what if something tries to update the value that was stored I read it" that I'm testing, but it's a little bit stronger than that. I'll pick better phrasing.